profiles/tests/test_profiles_graphql_gdpr_api.py

```diff
@@ -569,3 +585,66 @@ def test_api_tokens_missing(user_gql_client, service_1, query_or_delete, mocker)
         executed = user_gql_client.execute(DELETE_MY_PROFILE_MUTATION)
 
     assert_match_error_code(executed, MISSING_GDPR_API_TOKEN_ERROR)
+
+
+@pytest.mark.parametrize("dry_run", [True, False])
+def test_user_can_delete_data_from_a_service(
+    user_gql_client, service_1, service_2, mocker, dry_run
+):
+    query = DELETE_SERVICE_DATA_MUTATION_TEMPLATE.substitute(
+        service_name=service_1.name, dry_run=str(dry_run).lower()
+    )
+
+    def mock_delete_gdpr_data(self, api_token, dry_run=False):
+        return True
+
+    mocked_gdpr_delete = mocker.patch.object(
+        ServiceConnection,
+        "delete_gdpr_data",
+        autospec=True,
+        side_effect=mock_delete_gdpr_data,
+    )
+    mocker.patch.object(
+        TunnistamoTokenExchange, "fetch_api_tokens", return_value=GDPR_API_TOKENS
+    )
+    profile = ProfileFactory(user=user_gql_client.user)
+    ServiceConnectionFactory(profile=profile, service=service_1)
+    ServiceConnectionFactory(profile=profile, service=service_2)
+
+    user_gql_client.execute(query)
+
+    if dry_run:
+        assert mocked_gdpr_delete.call_count == 1
+        assert [c[2]["dry_run"] for c in mocked_gdpr_delete.mock_calls] == [True]
+        assert ServiceConnection.objects.count() == 2
+    else:
+        assert mocked_gdpr_delete.call_count == 2
+        assert [c[2]["dry_run"] for c in mocked_gdpr_delete.mock_calls] == [True, False]
+        assert ServiceConnection.objects.count() == 1
+
+
+def test_error_when_trying_to_delete_data_from_an_unknown_service(user_gql_client):
+    query = DELETE_SERVICE_DATA_MUTATION_TEMPLATE.substitute(
+        service_name="unknown_service", dry_run="false"
+    )
+
+    ProfileFactory(user=user_gql_client.user)
+
+    executed = user_gql_client.execute(query)
+
+    assert_match_error_code(executed, CONNECTED_SERVICE_DELETION_FAILED_ERROR)
```

Comment:
This test could have a Service, with a known name that is used in the query. Just that the Profile doesn't have a connection to that Service (the Profile could even have a connection to _another_ Service for completeness).

Reply:
Isn't that just setting superfluous things up?

Reply:
The service connection to another Service is probably unnecessary.

In [another comment](https://github.com/City-of-Helsinki/open-city-profile/pull/421#discussion_r984480600) I already suggested a dedicated error for the case where no service connection is found. But there are two reasons why a service connection may not be found:

1. Service with the given name doesn't even exist (this reason is checked in this test)
1. Service with the given name exists, but the Profile has no connection to it

Should these reasons even produce a different error code?

If we think about what the user should do next, I guess both cases are similar (or are they?): the user doesn't need to take any further action. That suggests that one error code for both reasons is enough.

Both reasons could still be tested.

Reply:
Added a test for the second case.